Add support for "from-image" to ImageBitmapOptions.

mattwoodrow opened the pull request:
#### 556e20c0b0053bc5c7f6d1a5635ad2467c6bade8
<pre>
Add support for &quot;from-image&quot; to ImageBitmapOptions.
<a href="https://bugs.webkit.org/show_bug.cgi?id=250476">https://bugs.webkit.org/show_bug.cgi?id=250476</a>

Reviewed by Tim Nguyen.

&quot;none&quot; is deprecated, and has been renamed to &quot;from-image&quot;.
This adds the new &quot;from-image&quot; value, with the same behaviour as &quot;none&quot;.

* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/manual/imagebitmap/createImageBitmap-flipY-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/manual/imagebitmap/createImageBitmap-sizeOverflow-expected.txt:
* Source/WebCore/html/ImageBitmapOptions.h:
* Source/WebCore/html/ImageBitmapOptions.idl:

Canonical link: <a href="https://commits.webkit.org/266893@main">https://commits.webkit.org/266893@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d152577e4aad280a9748b1dbc5f1a087878bc562

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14962 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15266 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15620 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16714 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14076 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15100 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17787 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15367 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16707 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15142 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15618 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12710 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17444 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12898 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13496 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20460 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13976 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13664 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16901 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14221 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12026 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13504 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3628 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17837 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14063 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/ImageBitmapOptions.h:
```diff
@@ -31,7 +31,7 @@
 
 struct ImageBitmapOptions {
-    enum class Orientation { None, FlipY };
+    enum class Orientation { None, FlipY, FromImage };
```
This line: https://searchfox.org/wubkat/rev/33c49ff64449389431294ba0ace6f7d0ad6306b7/Source/WebCore/html/ImageBitmap.cpp#890 has a check for equality to None. Does this need to be adjusted if the behavior is the same as "none"?

mattwoodrow replied:
Oh, good catch! It does (but it just breaks an optimisation without the adjustment, not correctness),

nt1m commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -887,7 +887,7 @@ void ImageBitmap::createCompletionHandler(ScriptExecutionContext& scriptExecutio
     // resulting ImageBuffer directly.
     auto alphaPremultiplication = alphaPremultiplicationForPremultiplyAlpha(options.premultiplyAlpha);
-    if (sourceRectangle.returnValue().location().isZero() && sourceRectangle.returnValue().size() == imageData->size() && sourceRectangle.returnValue().size() == outputSize && options.orientation == ImageBitmapOptions::Orientation::None) {
+    if (sourceRectangle.returnValue().location().isZero() && sourceRectangle.returnValue().size() == imageData->size() && sourceRectangle.returnValue().size() == outputSize && options.orientation != ImageBitmapOptions::Orientation::FlipY) {
```
Would `options.orientation == ImageBitmapOptions::Orientation::FromImage` be nicer here?

mattwoodrow replied:
Honestly, I think this code might just be wrong.

'None' previously meant 'use the image's orientation data', but this fast path is for when there's no rotation to be done (what you would expect 'None' to mean).

I think this should be checking `options.resolvedImageOrientation == None`.

mattwoodrow replied:
I take it back, this is the `ImageData` path, so it can't have its own orientation data.

Writing it as 'if we don't need to flip, then take the fast path' is nicer to me :)

nt1m replied:
ok sounds good!

mattwoodrow committed:
Add support for "from-image" to ImageBitmapOptions.
https://bugs.webkit.org/show_bug.cgi?id=250476

Reviewed by Tim Nguyen.

"none" is deprecated, and has been renamed to "from-image".
This adds the new "from-image" value, with the same behaviour as "none".

* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/manual/imagebitmap/createImageBitmap-flipY-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/canvas/element/manual/imagebitmap/createImageBitmap-sizeOverflow-expected.txt:
* Source/WebCore/html/ImageBitmapOptions.h:
* Source/WebCore/html/ImageBitmapOptions.idl:

Canonical link: https://commits.webkit.org/266893@main
